### PR TITLE
Implement NoInfer linter.

### PR DESCRIPTION
**Pull request by olafurpg**
This is the first scalafix linter rule. There are still a few rough corners around how to display carets around synthetics but the api is already usable. It was a great exercise to implement this linter since I found a few necessary parts missing in testkit to make the experience nice.

### CODE REVIEW

File: scalafix-testkit/src/main/scala/scalafix/testkit/SemanticRuleSuite.scala [hunk 1]
```diff
@@ -95,7 +96,7 @@ abstract class SemanticRuleSuite(
       val key = lintMessages.head.category.key(rule.name)
       val explanation =
         s"""|To fix this problem, suffix the culprit lines with
-            |   // scalafix: $key
+            |   // assert: $key
```
Comment (gabro) [inline scalafix-testkit/src/main/scala/scalafix/testkit/SemanticRuleSuite.scala]: how does the `// assert` comment work?

Comment (olafurpg) [reply scalafix-testkit/src/main/scala/scalafix/testkit/SemanticRuleSuite.scala]: You can assert which linter message is reported at which specfic lines https://github.com/scalacenter/scalafix/blob/master/scalafix-tests/input/src/main/scala/test/NoInfer.scala#L7

The test fails if
- the linter category doesn't match,
- a message is not reported on that line
- a message is reported at a line without an assert

Comment (gabro) [reply scalafix-testkit/src/main/scala/scalafix/testkit/SemanticRuleSuite.scala]: Nice! This is only for testkit right?

Comment (olafurpg) [reply scalafix-testkit/src/main/scala/scalafix/testkit/SemanticRuleSuite.scala]: Yes, this is only testkit. Escape hatches for users are yet unimplemented https://github.com/scalacenter/scalafix/issues/241

**Pull request merged**